src/tls13.rs

```diff
@@ -0,0 +1,313 @@
+
+        // Set up needed parameters for ChaCha decrypt
+        let nonce = Nonce::new(&self.iv, seq);
+        let auth_data = make_tls13_aad(payload_len); // The total message including tag and encoding byte must be used for auth data.
+        let mut tag = [0u8; GCM_TAG_LENGTH];
+        tag.copy_from_slice(&payload[message_length..]);
+
+        // Decrypting the payload in place, there is no +1 here since [`message_length`] accounts for the extra byte for encoding type.
+        match chacha20_poly1305_decrypt_in_place(
+            &self.key,
+            &nonce.0,
+            &auth_data,
+            &mut payload[..message_length],
+            &tag,
+        ) {
+            Ok(_) => {
+                payload.truncate(message_length);
+                msg.into_tls13_unpadded_message() // This removes the optional padding of zero bytes.
+            }
+            Err(symcrypt_error) => {
+                let custom_error_message = format!(
+                    "SymCryptError: {}",
+                    symcrypt_error.to_string() // Using general error to propagate the SymCrypt error back to the caller
+                );
+                return Err(rustls::Error::General(custom_error_message));
+            }
+        }
+    }
+}
+
+/// GCM for TLS 1.3
+
+/// Tls13Gcm impls [`Tls13AeadAlgorithm`].
+///
+/// [`algo_type`] represents either GCM128 or GCM256 which corresponds to a 16 and 32 byte key respectively.
+pub struct Tls13Gcm {
+    algo_type: AesGcm,
+}
+
+/// Algo types for GCM TLS 1.3 and TLS 1.2
+pub enum AesGcm {
+    Aes128Gcm,
+    Aes256Gcm,
+}
+
+impl AesGcm {
+    pub fn key_size(&self) -> usize {
+        match self {
+            AesGcm::Aes128Gcm => 16,
+            AesGcm::Aes256Gcm => 32,
+        }
+    }
+}
+
+/// Tls13GcmState impls [`MessageEncrypter`] and [`MessageDecrypter`]
+///
+/// [`key`] is a rust-symcrypt::GcmExpandedKey that has expands the provided key
+/// [`iv`] is an initialization vector that is needed to create the unique nonce.
+pub struct Tls13GcmState {
+    key: GcmExpandedKey,
+    iv: Iv,
+}
+
+impl Tls13AeadAlgorithm for Tls13Gcm {
+    fn encrypter(&self, key: AeadKey, iv: Iv) -> Box<dyn MessageEncrypter> {
+        // Unwrapping here, in the scenarios that GcmExpandKey would fail should result in a panic, ie: Not enough memory.
+        Box::new(Tls13GcmState {
+            key: GcmExpandedKey::new(key.as_ref(), BlockCipherType::AesBlock).unwrap(),
+            iv: iv,
+        })
+    }
+
+    fn decrypter(&self, key: AeadKey, iv: Iv) -> Box<dyn MessageDecrypter> {
+        // Unwrapping here, in the scenarios that GcmExpandKey would fail should result in a panic, ie: Not enough memory.
+        Box::new(Tls13GcmState {
+            key: GcmExpandedKey::new(key.as_ref(), BlockCipherType::AesBlock).unwrap(),
+            iv: iv,
+        })
+    }
+
+    fn key_len(&self) -> usize {
+        self.key_len()
```

Comment:
> self.key_len() [](http://example.com/codeflow?start=7&length=15)

This looks recursive??

Shouldn't it be self.algo_type.key_size()

Reply:
Changed